### ISSUE TITLE
give more functionality to pocketmine

### ISSUE THREAD

**Issue by royaljacques** [pmmp/PocketMine-MP#6257]
I've been using Pocketmine since version 2.0.0, and I've seen it evolve over the years.
I find it a shame that we don't follow the Java versions which always have the latest features, as well as the latest additions.
I don't understand why we absolutely don't want to do things correctly, and thus allow the developer to concentrate on something other than having to redo all the vanilla elements/blocks in Minecraft
I also don't understand the fact of having removed during the passage v3 => v4 => v5 the possibility of modifying the durability of an object or armor. This requires us to "cheat" to modify vanilla items in Minecraft. However, for some servers, being able to completely modify the element is a significant bonus compared to the rest of the code.

What I suggest:

- put back in place (as best as possible of course) the possibility of being able to modify certain statistics of vanilla items preloaded in pocketmine
- more easily accepted requests to add vanilla content from the community
- allow everyone, thanks to your great knowledge in minecraft php logic, to develop functionalities that are useful to everyone (I am thinking in particular of adding new non-vanilla items to the server like customies does). Doing it yourself would greatly optimize the servers, and reduce the number of plugins to load subsequently.
- I was talking about adding vanilla systems, I was thinking about dimension, which to this day are still not added in pocketmine. add portal functionality too. Why are these things, present since the beginning of Minecraft, still not present? why wait on plugins potentially published on poggit, which will never be correct and optimized?

**Comment by dktapps** [comment]
> I find it a shame that we don't follow the Java versions which always have the latest features, as well as the latest additions.

Java servers are just mods of the official server software, so they get all features for free from Mojang. PM is built from scratch and doesn't get the same benefit, as everything has to be re-implemented. This process is very time consuming, and there's a very small number of people in the community who are competent enough to write code fit for use on thousands of PM servers.

Our commitment has been to the quality of the existing server code, as opposed to cramming it chock full of poorly written code by merging every poor quality PR that comes our way. Accepting community contributions requires that those contributions are thoroughly reviewed for consistency, quality, and future maintainability. Very few PRs are perfect the first time around, but what often ends up happening is that PR authors end up abandoning their PRs as soon as the first review requesting changes comes in. The PR review process is also very slow and cumbersome, due to the long wait times between review <-> response, and nobody wants to do it because it's a time consuming and tedious process.

The bottom line PMMP doesn't have enough manpower or funding to create this utopia you're looking for. The community doesn't seem to care enough to properly fund the project, and expects me to write perfect code for them for nothing so that their servers can profit. (On the subject, I'll point out that the only reason PM5 has support for the newest Bedrock worlds is because Lifeboat sponsored the development of it. It was a huge task that I would not have bothered to undertake without their support.)

**Comment by pandaaaBE** [comment]
@dylan 

Honestly, you can't expect the first pull requests to be perfect - it takes time to familiarize yourself with the meta.

I don't know if you mean also my pull requests, but I didn't close mine because the first reviews suggested any changes, but because I simply didn't have time to work on them at the time.

So if you look at my PR then yes, what can I say, it was a mistake of mine that I opened so many and then had no more time for it.

But e.g. the Smithing Table PR is a great thing for many PocketMine servers but Himmel is still waiting for an answer.

**Comment by ShockedPlot7560** [comment]
We can't expect a first contribution to be perfect, and that wasn't my case either. However, greater commitment on the part of everyone involved in the review can help avoid inactive or very long PR.

As he said, PR is a tedious and time-consuming process that's no fun to do. The lack of staff at PMMP (probably indirectly linked to the lack of funding and commitment from the community) is a direct cause of this, drastically reducing treatment time.

And in your case, it's not the number of open PRs that's the problem, but rather the fact that it's boring to do (and, I don't think it's targeted at anyone, but more a general atmosphere within the project).

**Comment by ShockedPlot7560** [comment]
And in my opinion, instead of the community putting their energy into developing plugins to implement vanilla objects or logic, it would be more logical for them to put it into pocketmine to have your utopia thus said. There's certainly a lack of resources on the part of the pocketmine team and the community daring to review, but there's also a lack of will on the part of the community. 

Many people think that Dylan will do the job as he has done in recent years, relying on him and waiting for things to move quickly. However, to a person, it will always go slowly.

**Comment by dktapps** [comment]
> the fact that it's boring to do (and, I don't think it's targeted at anyone, but more a general atmosphere within the project).

this, and it takes so much back-and-forth (and reviewers spoonfeeding contributors) with such large delays that it's also quite demotivating. I've said before that many PRs could be done quicker if I were to reimplement them myself rather than reviewing other people's code, but I have time for neither of those things.

I've focused my energy on enhancing the API to expand what plugins are able to do, so that features don't need to be added to the PM core. This allows plugin developers to write stuff in plugins that may not be high enough quality to make it into the PM core, but still usable by users if they choose to (as well as giving users more options). In some cases these enhancements have had adverse side effects (such as the aforementioned inability to modify item damage rates in PM5), but that's mainly due to community reliance on anti-features that don't have a properly supported way to be done.

**Comment by dktapps** [comment]
The long and short of it is that we don't have the manpower or the funding to do all of the shiny things the community wants, so we have to focus our energy on the less exciting stuff that enables the community to build better plugins.

**Comment by SenseiTarzan** [comment]
> > I find it a shame that we don't follow the Java versions which always have the latest features, as well as the latest additions.
> 
> Java servers are just mods of the official server software, so they get all features for free from Mojang. PM is built from scratch and doesn't get the same benefit, as everything has to be re-implemented. This process is very time consuming, and there's a very small number of people in the community who are competent enough to write code fit for use on thousands of PM servers.
> 
> Our commitment has been to the quality of the existing server code, as opposed to cramming it chock full of poorly written code by merging every poor quality PR that comes our way. Accepting community contributions requires that those contributions are thoroughly reviewed for consistency, quality, and future maintainability. Very few PRs are perfect the first time around, but what often ends up happening is that PR authors end up abandoning their PRs as soon as the first review requesting changes comes in. The PR review process is also very slow and cumbersome, due to the long wait times between review <-> response, and nobody wants to do it because it's a time consuming and tedious process.
> 
> The bottom line PMMP doesn't have enough manpower or funding to create this utopia you're looking for. The community doesn't seem to care enough to properly fund the project, and expects me to write perfect code for them for nothing so that their servers can profit. (On the subject, I'll point out that the only reason PM5 has support for the newest Bedrock worlds is because Lifeboat sponsored the development of it. It was a huge task that I would not have bothered to undertake without their support.)

Why do plugins have to manipulate the API to be able to replace vanilla blocks when if you had left the code people would be happy because they replace the behavior of the blocks a lot , you forgot that this was an essential point now that it is no longer possible many no longer want to continue developing block vanilla plugins whereas at the time there was teaspoon Even if it was not perfect it pleased the community and now it is no longer possible to make redstone because you remove it
